Churn handling (#67)
c4dt/arti-rest: tools/README.md

# Tools for generating custom directory information

To reduce bandwidth use of lightarti-rest, the library relies on custom Tor directory information. The scripts in the `tools/` directory generate the required files, that can then be downloaded by apps for use with lightarti-rest.

> The scripts in this directory aim to pick reliable nodes so that ideally the consensus can be used for up to a week. We refer to the [accompanying Jupyter notebook](churn_analysis.ipynb) for a historical analysis of churn on the Tor network that shows shows that this approach is viable.

Reviewer: ```suggestion
The scripts in this directory pick reliable nodes so that ideally the consensus can be used for up to a week. Reliability is measured as INSERT MEASURES HERE. We refer to the [accompanying Jupyter notebook](churn_analysis.ipynb) for a historical analysis of churn on the Tor network that shows shows that this approach is viable.
```

The tool provides three functions:

> 1. Setting up a custom directory authority. We will use the keys of this authority to sign the smaller custom consensus files. This step only needs to be run once.

Reviewer: ```suggestion
1. Setting up a custom directory authority. The keys of this authority are used to sign the smaller custom consensus files. This step only needs to be run once.
```

2. Generating a new custom consensus. Based on the latest consensus information, the script produces a smaller consensus of reliable nodes.  A new custom consensus should ideally be generated at least once a day. Generating a new custom consensus requires the keys from the previous step.
3. Generating a churn file. As a custom consensus ages, some nodes may no longer be available. To increase efficiency, lightarti-rest can take as input a very small churn file that lists available nodes. A churn file is generated with respect to a specific custom consensus. Ideally it is updated every hour.

> We describe these in more detail below.

Reviewer: Can you add an example of size of the consensus and the churn file? I think it's 1MB and 10kB. This is interesting for readers.

Author: Good point, I am going to add some reasonable measurements. Essentially something like this:

The size of a consensus retrieved from the tor authorities in microconsensus format is about 2.2 MB.

The size of a shortened consensus will depend on the number of relay we decide to include, if we keep the default values (120 routers), it will have a size of about 45 KB.

The size of the churn file will increase over time, as more relays become unreachable. Each unreachable relay increase the size of the churn file by 41 bytes (40 bytes for the fingerprint of the relay plus 1 byte for the new line). So typically, its size will be less 1 KB.

Reviewer: That should go in the README file...

  `microdescriptors.txt`)

## Generate Churn

To improve the reliability of the subset of the Tor network upon which
Lightarti-rest relies to build circuits, we are providing a way to compute
a list of no longer available/reachable routers.

> This list of churned routers is intended to be a lightweight file containing

Reviewer: ```suggestion
This list of unavailable routers is intended to be a lightweight file containing
```
